osquery/tables/system/windows/ntfs_acl_permissions.cpp

```diff
@@ -0,0 +1,227 @@
+            TLOG << "LookupAccountSid error: " << GetLastError();
+		}
+		else {
+			return name;
+		}
+	case TRUSTEE_IS_OBJECTS_AND_NAME:
+		// ptstrName member is a pointer to an OBJECTS_AND_NAME struct
+		return ((OBJECTS_AND_NAME_*) t.ptstrName)->ptstrName;
+	default:
+		return "";
+	}
+}
+
+QueryData genNTFSACLPerms(QueryContext& context) {
+  QueryData results;
+
+	std::string ObjName;
+	unsigned long result = 0;
+	PACL DACL = NULL;
+	PEXPLICIT_ACCESS aceList = NULL, aceItem;
+	unsigned long aceCount, aceIndex;
+
+	auto paths = context.constraints["path"].getAll(EQUALS);
+    for (const auto& path_string : paths) {
+          if (fs::exists(path_string)) {
+				fs::path ObjName = path_string;
+
+				// Get a pointer to the existing DACL.
+
+                result = GetNamedSecurityInfo(path_string.c_str(),
+                                                SE_FILE_OBJECT,
+                                                DACL_SECURITY_INFORMATION,
+                                                NULL,
+                                                NULL,
+                                                &DACL,
+                                                NULL,
+                                                NULL);
+                if (ERROR_SUCCESS != result) {
+                    //printf("GetNamedSecurityInfo Error %u\n", result);
+					TLOG << "GetExplicitEnteriesFromAcl Error " << result;
+                }
+
+				// get list of ACEs from DACL pointer
+                result = GetExplicitEntriesFromAcl(
+                    DACL, &aceCount, &aceList);
+                if (ERROR_SUCCESS != result) {
+                    //printf("GetExplicitEnteriesFromAcl Error %u\n",
+                    //        result);
+					TLOG << "GetExplicitEnteriesFromAcl Error " << result;
+                }
+
+                // Loop through list of entries
+
+                aceItem = aceList;
+                // printf("AccessMode | TrusteeName | AccessPermissions
+                // | Inheritence\n");
+                for (aceIndex = 0; aceIndex < aceCount; aceItem++) {
+                    Row r;
+                    std::string accessMode;
+                    std::string inheritedFrom;
+                    std::string perms;
+                    std::string trusteeName;
+
+                    perms =
+                        accessPermsToStr(aceItem->grfAccessPermissions);
+                    accessMode = accessModeToStr(aceItem->grfAccessMode);
+                    inheritedFrom =
+                        inheritanceToStr(aceItem->grfInheritance);
+                    trusteeName = trusteeToStr(aceItem->Trustee);
+
+                    r["path"] = TEXT(path_string);
+                    r["type"] = TEXT(accessMode);
+                    r["principal"] = TEXT(trusteeName);
+                    r["access"] = TEXT(perms);
+                    r["inherited_from"] = TEXT(inheritedFrom);
+                    results.push_back(r);
```

Comment:
to avoid useless copy of `Row` object, you can move `r` into `push_back` here.

Reply:
Could you give a code example of what you mean?

Reply:
Yes, sure. Look:

```c++
results.push_back(std::move(r));
```
You can read about more semantics [here](http://en.cppreference.com/w/cpp/concept/MoveAssignable)